File: appletv/Sources/HexStackerKit/Net/RoomFlow.swift
import Foundation

/// A player slot in the room. A reference type so game-owned fields can be
/// mutated in place and seen by both RoomFlow and the coordinator, mirroring
/// the shared-record aliasing in the JS implementation.
public final class PlayerRecord {
    public let peerIndex: Int
    public var joinedAt: Int           // monotonic counter (not wall clock)
    public var connected: Bool
    // Game-owned fields (RoomFlow never reads these for its own logic):
    public var playerName: String
    public var colorSlot: Int          // dense 0..MAX_PLAYERS-1 (JS: playerIndex)
    public var startLevel: Int         // 1...15

    init(peerIndex: Int, joinedAt: Int, playerName: String, colorSlot: Int, startLevel: Int) {
        self.peerIndex = peerIndex
        self.joinedAt = joinedAt
        self.connected = true
        self.playerName = playerName
        self.colorSlot = colorSlot
        self.startLevel = startLevel
    }
}

/// Headless room / lobby / host state machine. Ported from partyplug/RoomFlow.js.
/// Owns identity, presence and host election; knows nothing about rendering or
/// the relay. The display (coordinator) drives transitions and reads `host`.
public final class RoomFlow {

    // Event callbacks (the kit's tiny pub/sub, as closures).
    public var onStateChange: ((_ from: RoomState, _ to: RoomState) -> Void)?
    public var onPlayerJoin: ((PlayerRecord) -> Void)?
    public var onPlayerUpdate: ((PlayerRecord) -> Void)?
    public var onPlayerLeave: ((_ peerIndex: Int) -> Void)?
    public var onRosterChange: ((_ players: [PlayerRecord]) -> Void)?
    public var onHostChange: ((_ hostPeerIndex: Int?) -> Void)?

    /// Platform master provider (AirConsole). Returns nil in relay mode.
    public var masterProvider: (() -> Int?)?

    public private(set) var state: RoomState = .lobby
    /// Raw sticky host slot (may differ from effective `host` during a blip).
    public private(set) var hostPeerIndex: Int?

    private var players: [Int: PlayerRecord] = [:]
    private var disconnected: Set<Int> = []
    private var order: [Int] = []
    private var joinSeq = 0

    // Liveness (presence-timeout) + late-joiner grace. Clock-free: every method is
    // a pure predicate the host calls with an injected nowMs (mirrors the canonical
    // partyplug/RoomFlow.js). Detectors never mutate `disconnected` — the host
    // applies a detected expiry through markDisconnected, preserving single-writer.
    private var lastSeen: [Int: Double] = [:]
    private let livenessTimeoutMs: Double
    private let graceMs: Double
    private var graceDeadline: Double?

    public init(livenessTimeoutMs: Double = 3000, graceMs: Double = 5000) {
        self.livenessTimeoutMs = livenessTimeoutMs
        self.graceMs = graceMs
    }

    private static let validTransitions: [RoomState: [RoomState]] = [
        .lobby: [.countdown],
        .countdown: [.playing, .lobby],
        .playing: [.results, .lobby],
        .results: [.countdown, .lobby],
    ]

    // MARK: - Transitions

    @discardableResult
    public func transition(to: RoomState) -> Bool {
        let from = state
        if to == from { return true }
        guard Self.validTransitions[from]?.contains(to) == true else {
            return false   // invalid transition (JS warns and returns false)
        }
        emittingHostChange {
            state = to
            if to == .countdown { snapshotOrder() }
            if to == .lobby { order = [] }
            if to != .playing { graceDeadline = nil }   // a stale deadline must not fire in the next game
            if to == .lobby || to == .results { reconcileStickyHost() }
        }
        onStateChange?(from, to)
        return true
    }

    public func endGame() { transition(to: .results) }
    public func returnToLobby() { transition(to: .lobby) }

    // MARK: - Roster

    /// Add a new player or refresh an existing one (reconnect). Returns the record.
    @discardableResult
    public func addPlayer(peerIndex: Int, playerName: String, colorSlot: Int, startLevel: Int = 1) -> PlayerRecord {
        if let existing = players[peerIndex] {
            // Mutate inside the wrapper so a host-eligibility change (a sticky
            // host regaining `connected`) actually emits (JS captures prevHost
            // before the mutations).
            emittingHostChange {
                existing.playerName = playerName
                existing.colorSlot = colorSlot
                existing.startLevel = startLevel
                existing.connected = true            // joinedAt preserved
                disconnected.remove(peerIndex)
            }
            onPlayerUpdate?(existing)
            onRosterChange?(list())
            return existing
        }
        let rec = PlayerRecord(peerIndex: peerIndex, joinedAt: joinSeq, playerName: playerName,
                               colorSlot: colorSlot, startLevel: startLevel)
        joinSeq += 1
        players[peerIndex] = rec
        // First joiner owns the sticky host slot.
        if hostPeerIndex == nil {
            hostPeerIndex = peerIndex
            onHostChange?(host)
        }
        onPlayerJoin?(rec)
        onRosterChange?(list())
        return rec
    }

    public func removePlayer(_ peerIndex: Int) {
        guard players[peerIndex] != nil else { return }
        emittingHostChange {
            let wasHost = peerIndex == hostPeerIndex
            players.removeValue(forKey: peerIndex)
            disconnected.remove(peerIndex)
            lastSeen.removeValue(forKey: peerIndex)
            order.removeAll { $0 == peerIndex }
            // Sticky slot only moves on departure in LOBBY/RESULTS.
            if wasHost, state == .lobby || state == .results {
                hostPeerIndex = electNextHost(exclude: peerIndex)
            }
        }
        onPlayerLeave?(peerIndex)
        onRosterChange?(list())
    }

    /// Cross-device takeover: a different client claims a dropped slot under a
    /// new peer index. Returns false if not applicable.
    @discardableResult
    public func rekey(oldId: Int, newId: Int) -> Bool {
        guard oldId != newId, let rec = players[oldId] else { return false }
        emittingHostChange {
            players.removeValue(forKey: oldId)
            players.removeValue(forKey: newId)   // drop placeholder slot
            // PlayerRecord.peerIndex is immutable; create a moved copy.
            let moved = PlayerRecord(peerIndex: newId, joinedAt: rec.joinedAt, playerName: rec.playerName,
                                     colorSlot: rec.colorSlot, startLevel: rec.startLevel)
            moved.connected = true
            players[newId] = moved
            disconnected.remove(oldId)
            disconnected.remove(newId)
            // Keep the NEWER of the two last-seen stamps (mirrors partyplug/RoomFlow.js):
            // the placeholder's stamp is the claimant's live signal; the old seat's is up
            // to a reconnect-grace window stale and survives only as a fallback for a
            // never-seen placeholder.
            let keptStamp = [lastSeen.removeValue(forKey: oldId), lastSeen.removeValue(forKey: newId)]
                .compactMap { $0 }.max()
            if let keptStamp { lastSeen[newId] = keptStamp }
            order = order.map { $0 == oldId ? newId : $0 }
            if hostPeerIndex == oldId { hostPeerIndex = newId }
        }
        onRosterChange?(list())
        return true
    }

    public func markDisconnected(_ peerIndex: Int) {
        guard let p = players[peerIndex] else { return }
        emittingHostChange {
            p.connected = false
            disconnected.insert(peerIndex)
        }
        onRosterChange?(list())
    }

    public func markReconnected(_ peerIndex: Int) {
        guard let p = players[peerIndex] else { return }
        emittingHostChange {
            p.connected = true
            disconnected.remove(peerIndex)
        }
        onRosterChange?(list())
    }

    public func clearDisconnected() {
        guard !disconnected.isEmpty else { return }
        emittingHostChange {
            disconnected.removeAll()
            for (_, p) in players { p.connected = true }
        }
        onRosterChange?(list())
    }

    /// Replace the active participant order (the game's player order) so host
    /// eligibility is restricted to current participants during a round.
    public func setActiveOrder(_ peerIndices: [Int]) {
        emittingHostChange {
            order = peerIndices.filter { players[$0] != nil }
        }
    }

    public func reset() {
        let hadHost = hostPeerIndex != nil
        let prevState = state
        players.removeAll()
        disconnected.removeAll()
        lastSeen.removeAll()
        graceDeadline = nil
        order = []
        hostPeerIndex = nil
        joinSeq = 0
        state = .lobby
        if prevState != .lobby { onStateChange?(prevState, .lobby) }
        onRosterChange?([])
        if hadHost { onHostChange?(nil) }
    }

    // MARK: - Queries

    public var size: Int { players.count }
    public var connectedCount: Int { players.count - disconnected.count }
    public func player(_ peerIndex: Int) -> PlayerRecord? { players[peerIndex] }
    public func contains(_ peerIndex: Int) -> Bool { players[peerIndex] != nil }
    public func isDisconnected(_ peerIndex: Int) -> Bool { disconnected.contains(peerIndex) }

    // MARK: - Liveness / presence timeout (ported from partyplug/RoomFlow.js)

    /// Record that we just heard from a peer (a controller message / rejoin).
    /// Ignores unknown peers so a stray packet can't resurrect a stamp.
    public func onSeen(_ peerIndex: Int, _ nowMs: Double) {
        if players[peerIndex] != nil { lastSeen[peerIndex] = nowMs }
    }

    /// Re-stamp every current player as just-seen. Called on the "everyone present"
    /// transitions (game start) so a controller that went briefly quiet isn't
    /// instantly flagged once the LOBBY gate lifts in COUNTDOWN/PLAYING.
    public func primeLiveness(_ nowMs: Double) {
        for id in players.keys { lastSeen[id] = nowMs }
    }

    /// Has this peer gone silent past the liveness window? Strict `>`: exactly-at
    /// timeout is still alive. A peer with no stamp (joined but never messaged) is
    /// always-alive here; its only cleanup path is peer_left.
    public func isExpired(_ peerIndex: Int, _ nowMs: Double) -> Bool {
        guard let seen = lastSeen[peerIndex] else { return false }
        return nowMs - seen > livenessTimeoutMs
    }

    /// Peers that just crossed the liveness window and aren't already flagged
    /// disconnected. Empty in LOBBY (a silent idle controller there is fine).
    public func expiredPeers(_ nowMs: Double) -> [Int] {
        guard state != .lobby else { return [] }
        return players.keys.filter { !disconnected.contains($0) && isExpired($0, nowMs) }.sorted()
    }

    /// Is every active participant (the `order`) currently disconnected? False when
    /// there is no active order (lobby), so an empty game never auto-pauses.
    public var allParticipantsDisconnected: Bool {
        guard !order.isEmpty else { return false }
        return !order.contains { !disconnected.contains($0) }
    }

    /// Any roster member not in the active participant order (joined after the game
    /// started). They're who the grace window waits for.
    public var hasLateJoiners: Bool {
        players.keys.contains { !order.contains($0) }
    }

    /// Deadline-driven late-joiner grace. While PLAYING with every participant gone
    /// but late joiners waiting, the first call arms a deadline at nowMs + graceMs
    /// and returns false; a later call returns true exactly once when it elapses.
    /// Any frame where the condition no longer holds clears the deadline, so a
    /// reconnect implicitly cancels the return-to-lobby.
    public func graceTick(_ nowMs: Double) -> Bool {
        if state == .playing && allParticipantsDisconnected && hasLateJoiners {
            if graceDeadline == nil { graceDeadline = nowMs + graceMs; return false }
            if nowMs >= graceDeadline! { graceDeadline = nil; return true }
            return false
        }
        graceDeadline = nil
        return false
    }

    /// Roster sorted by join order.
    public func list() -> [PlayerRecord] {
        players.values.sorted { $0.joinedAt < $1.joinedAt }
    }

    /// Lowest free color slot (0..MAX_PLAYERS-1), or -1 if the room is full.
    public func lowestFreeSlot() -> Int {
        let taken = Set(players.values.map { $0.colorSlot })
        for slot in 0..<EngineConstants.maxPlayers where !taken.contains(slot) { return slot }
        return -1
    }

    public func takenColorSlots() -> [Int] {
        players.values.map { $0.colorSlot }.sorted()
    }

    /// The effective host: platform master, else sticky slot, else oldest-joined
    /// eligible present player.
    public var host: Int? {
        let eligible: Set<Int>? = restricted ? Set(order) : nil
        if let provider = masterProvider, let m = provider(), isEligible(m, eligible) { return m }
        if isEligible(hostPeerIndex, eligible) { return hostPeerIndex }
        return oldestEligible(eligible, exclude: nil)
    }

    // MARK: - Internals

    private var restricted: Bool {
        (state == .countdown || state == .playing || state == .results) && !order.isEmpty
    }

    private func isEligible(_ idx: Int?, _ set: Set<Int>?) -> Bool {
        guard let idx, players[idx] != nil, !disconnected.contains(idx) else { return false }
        if let set { return set.contains(idx) }
        return true
    }

    private func oldestEligible(_ set: Set<Int>?, exclude: Int?) -> Int? {
        var best: Int?
        var bestJoin = Int.max
        for (idx, rec) in players {
            if idx == exclude { continue }
            guard isEligible(idx, set) else { continue }
            if rec.joinedAt < bestJoin { bestJoin = rec.joinedAt; best = idx }
        }
        return best
    }

    private func electNextHost(exclude: Int?) -> Int? {
        oldestEligible(restricted ? Set(order) : nil, exclude: exclude)
    }

    private func reconcileStickyHost() {
        // An empty room keeps the slot untouched (JS parity); nulling it here
        // would emit a spurious host change on an empty-room state transition.
        guard !players.isEmpty else { return }
        let eligible: Set<Int>? = restricted ? Set(order) : nil
        if isEligible(hostPeerIndex, eligible) { return }
        hostPeerIndex = electNextHost(exclude: hostPeerIndex)
    }

    private func snapshotOrder() {
        order = players.values
            .filter { $0.connected }
            .sorted { $0.joinedAt < $1.joinedAt }
            .map { $0.peerIndex }
    }

    /// Runs `body`, emitting `onHostChange` if the effective host changed.
    private func emittingHostChange(_ body: () -> Void) {
        let before = host
        body()
        let after = host
        if before != after { onHostChange?(after) }
    }
}

File: android/core/src/commonMain/kotlin/com/hexstacker/core/room/RoomFlow.kt
package com.hexstacker.core.room

import com.hexstacker.core.net.RoomState

/**
 * Headless room / lobby / host state machine. A faithful Kotlin port of
 * `partyplug/RoomFlow.js` (the canonical, battle-tested implementation, kept as
 * the parity oracle in `RoomFlowParityTest`). Owns identity, join order,
 * presence and host election; knows nothing about rendering, the relay, or game
 * fields like color/name/score (those are opaque data on the live record).
 *
 * Pure, synchronous, single-writer: no clock reads (liveness takes an injected
 * `nowMs`), no transport, no rendering. Callbacks are Kotlin lambdas (the JS
 * event emitter ported to closures, as the Swift port did).
 *
 * This port follows the JS shape exactly (per-op `prevHost` capture + emit)
 * rather than the Swift `emittingHostChange` wrapper, so the parity oracle holds
 * byte-for-byte.
 */
class RoomFlow(
    /** Platform master (AirConsole). Returns null in relay mode; null disables. */
    private val masterProvider: (() -> Int?)? = null,
    livenessTimeoutMs: Double = Double.POSITIVE_INFINITY,
    graceMs: Double = 0.0,
    /** When it returns false, liveness expiry is suppressed. Read live. */
    private val livenessEnabledProvider: (() -> Boolean)? = null,
) {
    // ---- callbacks (assign as needed; all optional) ----
    var onStateChange: ((from: RoomState, to: RoomState) -> Unit)? = null
    var onPlayerJoin: ((PlayerRecord) -> Unit)? = null
    var onPlayerUpdate: ((PlayerRecord) -> Unit)? = null
    var onPlayerLeave: ((peerIndex: Int) -> Unit)? = null
    var onRosterChange: ((players: List<PlayerRecord>) -> Unit)? = null
    var onHostChange: ((hostPeerIndex: Int?) -> Unit)? = null

    // ---- internal state (mirrors RoomFlow.js exactly) ----
    private val players = LinkedHashMap<Int, PlayerRecord>() // insertion-ordered for stable iteration
    private val disconnected = HashSet<Int>()
    private var order = mutableListOf<Int>()                 // active participant peerIndices
    private var joinSeq = 0
    private var hostSlot: Int? = null                        // backing for hostPeerIndex
    private var roomState = RoomState.LOBBY                  // backing for state

    // liveness
    private val lastSeen = HashMap<Int, Double>()
    private val livenessTimeoutMs = livenessTimeoutMs
    private val graceMs = graceMs
    private var graceDeadline: Double? = null

    // ---- observable state ----
    val state: RoomState get() = roomState
    val hostPeerIndex: Int? get() = hostSlot
    val size: Int get() = players.size
    val connectedCount: Int get() = players.size - disconnected.size

    /**
     * Effective host: platform master (if eligible) -> sticky slot (if eligible)
     * -> oldest-joined eligible present player. Read-only; the sticky slot is only
     * mutated by [removePlayer] / [rekey] / [reconcileStickyHost].
     */
    val host: Int?
        get() {
            val eligible = if (restricted()) order.toHashSet() else null
            masterProvider?.let { mp ->
                val m = mp()
                if (isEligible(m, eligible)) return m
            }
            if (isEligible(hostSlot, eligible)) return hostSlot
            return oldestEligible(eligible, exclude = null)
        }

    companion object {
        /** server/constants.js MAX_PLAYERS. */
        const val MAX_PLAYERS = 8

        /** Lowest free dense slot in [0, max); pass slot values, NOT peerIndices. */
        fun lowestFreeSlot(used: Iterable<Int>, max: Int): Int {
            val taken = used.toHashSet()
            for (i in 0 until max) if (i !in taken) return i
            return -1
        }

        private val VALID_TRANSITIONS: Map<RoomState, List<RoomState>> = mapOf(
            RoomState.LOBBY to listOf(RoomState.COUNTDOWN),
            RoomState.COUNTDOWN to listOf(RoomState.PLAYING, RoomState.LOBBY),
            RoomState.PLAYING to listOf(RoomState.RESULTS, RoomState.LOBBY),
            RoomState.RESULTS to listOf(RoomState.COUNTDOWN, RoomState.LOBBY),
        )
    }

    // =====================================================================
    // Roster
    // =====================================================================

    /** Add a new player or reconnect/refresh an existing one (same peerIndex). */
    fun addPlayer(peerIndex: Int, playerName: String, colorSlot: Int, startLevel: Int = 1): PlayerRecord {
        val existing = players[peerIndex]
        if (existing != null) {
            val prevHost = host
            // Reconnect: refresh game fields + presence; joinedAt preserved.
            existing.playerName = playerName
            existing.colorSlot = colorSlot
            existing.startLevel = startLevel
            existing.connected = true
            disconnected.remove(peerIndex)
            if (host != prevHost) onHostChange?.invoke(host)
            onPlayerUpdate?.invoke(existing)
            onRosterChange?.invoke(list())
            return existing
        }
        val rec = PlayerRecord(peerIndex, joinSeq++, playerName, colorSlot, startLevel)
        rec.connected = true
        players[peerIndex] = rec
        // First joiner owns the sticky host slot (also covers "room emptied then rejoined").
        if (hostSlot == null) {
            hostSlot = peerIndex
            onHostChange?.invoke(host)
        }
        onPlayerJoin?.invoke(rec)
        onRosterChange?.invoke(list())
        return rec
    }

    /** Hard leave (peer_left). Sticky slot only moves on departure in LOBBY/RESULTS. */
    fun removePlayer(peerIndex: Int) {
        if (!players.containsKey(peerIndex)) return
        val prevHost = host
        val wasHost = peerIndex == hostSlot
        players.remove(peerIndex)
        disconnected.remove(peerIndex)
        lastSeen.remove(peerIndex)
        order.remove(peerIndex)
        if (wasHost && (roomState == RoomState.LOBBY || roomState == RoomState.RESULTS)) {
            hostSlot = electNextHost(peerIndex)
        }
        if (host != prevHost) onHostChange?.invoke(host)
        onPlayerLeave?.invoke(peerIndex)
        onRosterChange?.invoke(list())
    }

    /** Cross-device takeover: a different client claims a dropped slot under a new id. */
    fun rekey(oldId: Int, newId: Int): Boolean {
        if (oldId == newId) return false
        val rec = players[oldId] ?: return false
        val prevHost = host
        players.remove(oldId)
        players.remove(newId) // drop the placeholder slot the returning peer got
        // PlayerRecord.peerIndex is val -> create a moved copy preserving joinedAt + game fields.
        val moved = PlayerRecord(newId, rec.joinedAt, rec.playerName, rec.colorSlot, rec.startLevel)
        moved.connected = true
        moved.lastPingTime = rec.lastPingTime
        players[newId] = moved
        disconnected.remove(oldId)
        disconnected.remove(newId)
        // Keep the NEWER of the two last-seen stamps (mirrors partyplug/RoomFlow.js):
        // the placeholder's stamp is the claimant's live signal; the old seat's is up
        // to a reconnect-grace window stale and survives only as a fallback for a
        // never-seen placeholder.
        val keptStamp = listOfNotNull(lastSeen.remove(oldId), lastSeen.remove(newId)).maxOrNull()
        if (keptStamp != null) lastSeen[newId] = keptStamp
        for (i in order.indices) if (order[i] == oldId) order[i] = newId
        if (hostSlot == oldId) hostSlot = newId
        if (host != prevHost) onHostChange?.invoke(host)
        onRosterChange?.invoke(list())
        return true
    }

    /** Soft disconnect window (record stays; presence flips false). */
    fun markDisconnected(peerIndex: Int) {
        val p = players[peerIndex] ?: return
        val prevHost = host
        p.connected = false
        disconnected.add(peerIndex)
        if (host != prevHost) onHostChange?.invoke(host)
        onRosterChange?.invoke(list())
    }

    fun markReconnected(peerIndex: Int) {
        val p = players[peerIndex] ?: return
        val prevHost = host
        p.connected = true
        disconnected.remove(peerIndex)
        if (host != prevHost) onHostChange?.invoke(host)
        onRosterChange?.invoke(list())
    }

    /** Clear every disconnect flag, marking all current players present. */
    fun clearDisconnected(nowMs: Double? = null) {
        // Re-stamp liveness on this "everyone present" transition first (unconditional when nowMs given).
        if (nowMs != null) for (id in players.keys) lastSeen[id] = nowMs
        if (disconnected.isEmpty()) return
        val prevHost = host
        disconnected.clear()
        for (p in players.values) p.connected = true
        if (host != prevHost) onHostChange?.invoke(host)
        onRosterChange?.invoke(list())
    }

    /** Replace the active participant order so host eligibility tracks the game's order. */
    fun setActiveOrder(peerIndices: List<Int>) {
        order = peerIndices.filter { players.containsKey(it) }.toMutableList()
    }

    /** Reset to a fresh room (new room / return to welcome). */
    fun reset() {
        val prevState = roomState
        val hadHost = hostSlot != null
        players.clear()
        disconnected.clear()
        lastSeen.clear()
        graceDeadline = null
        order = mutableListOf()
        hostSlot = null
        joinSeq = 0
        roomState = RoomState.LOBBY
        if (prevState != RoomState.LOBBY) onStateChange?.invoke(prevState, RoomState.LOBBY)
        onRosterChange?.invoke(emptyList())
        if (hadHost) onHostChange?.invoke(null)
    }

    // =====================================================================
    // Lifecycle
    // =====================================================================

    /** Validated state transition. Returns true if applied. */
    fun transitionTo(to: RoomState): Boolean {
        val from = roomState
        if (to == from) return true
        val allowed = VALID_TRANSITIONS[from]
        if (allowed == null || !allowed.contains(to)) return false
        roomState = to
        if (to != RoomState.PLAYING) graceDeadline = null
        if (to == RoomState.COUNTDOWN) snapshotOrder()
        if (to == RoomState.LOBBY) order = mutableListOf()
        if (to == RoomState.LOBBY || to == RoomState.RESULTS) reconcileStickyHost()
        onStateChange?.invoke(from, to)
        return true
    }

    fun endGame(): Boolean = transitionTo(RoomState.RESULTS)
    fun returnToLobby(): Boolean = transitionTo(RoomState.LOBBY)

    // =====================================================================
    // Queries
    // =====================================================================

    fun player(peerIndex: Int): PlayerRecord? = players[peerIndex]
    fun contains(peerIndex: Int): Boolean = players.containsKey(peerIndex)
    fun isDisconnected(peerIndex: Int): Boolean = disconnected.contains(peerIndex)
    fun isHost(peerIndex: Int): Boolean = peerIndex == host

    /** Roster sorted by join order. */
    fun list(): List<PlayerRecord> = players.values.sortedBy { it.joinedAt }

    /** Lowest free color slot (0..MAX_PLAYERS-1), or -1 if the room is full. */
    fun lowestFreeSlot(): Int = lowestFreeSlot(players.values.map { it.colorSlot }, MAX_PLAYERS)

    fun takenColorSlots(): List<Int> = players.values.map { it.colorSlot }.sorted()

    // =====================================================================
    // Host election (DisplayState.getHostPeerIndex / electNextHost / reconcile)
    // =====================================================================

    private fun restricted(): Boolean =
        (roomState == RoomState.COUNTDOWN || roomState == RoomState.PLAYING || roomState == RoomState.RESULTS) &&
            order.isNotEmpty()

    private fun isEligible(idx: Int?, eligibleSet: Set<Int>?): Boolean =
        idx != null && players.containsKey(idx) && !disconnected.contains(idx) &&
            (eligibleSet == null || eligibleSet.contains(idx))

    /** Oldest-joined present player within eligibleSet (null = everyone present), skipping excludeId. */
    private fun oldestEligible(eligibleSet: Set<Int>?, exclude: Int?): Int? {
        var best: Int? = null
        var bestJoin = Int.MAX_VALUE
        for ((id, rec) in players) {
            if (id == exclude) continue
            if (disconnected.contains(id)) continue
            if (eligibleSet != null && !eligibleSet.contains(id)) continue
            if (rec.joinedAt < bestJoin) {
                bestJoin = rec.joinedAt
                best = id
            }
        }
        return best
    }

    private fun electNextHost(exclude: Int?): Int? =
        oldestEligible(if (restricted()) order.toHashSet() else null, exclude)

    private fun reconcileStickyHost() {
        if (players.isEmpty()) return
        val eligible = if (restricted()) order.toHashSet() else null
        val current = hostSlot
        if (current != null && players.containsKey(current) && !disconnected.contains(current) &&
            (eligible == null || eligible.contains(current))
        ) {
            return
        }
        hostSlot = electNextHost(current)
        if (hostSlot != current) onHostChange?.invoke(host)
    }

    private fun snapshotOrder() {
        order = players.values
            .filter { it.connected }
            .sortedBy { it.joinedAt }
            .map { it.peerIndex }
            .toMutableList()
    }

    // =====================================================================
    // Liveness (pure, nowMs-injected predicates; never mutate disconnected, never emit)
    // =====================================================================

    fun onSeen(peerIndex: Int, nowMs: Double) {
        if (players.containsKey(peerIndex)) lastSeen[peerIndex] = nowMs
    }

    fun isExpired(peerIndex: Int, nowMs: Double): Boolean {
        if (livenessEnabledProvider?.invoke() == false) return false
        val seen = lastSeen[peerIndex] ?: return false
        return nowMs - seen > livenessTimeoutMs
    }

    fun expiredPeers(nowMs: Double): List<Int> {
        if (roomState == RoomState.LOBBY) return emptyList()
        val out = mutableListOf<Int>()
        for (id in players.keys) {
            if (disconnected.contains(id)) continue
            if (isExpired(id, nowMs)) out.add(id)
        }
        return out
    }

    fun allParticipantsDisconnected(): Boolean {
        if (order.isEmpty()) return false
        return order.all { disconnected.contains(it) }
    }

    fun hasLateJoiners(): Boolean = players.keys.any { !order.contains(it) }

    fun graceTick(nowMs: Double): Boolean {
        if (roomState == RoomState.PLAYING && allParticipantsDisconnected() && hasLateJoiners()) {
            val deadline = graceDeadline
            if (deadline == null) {
                graceDeadline = nowMs + graceMs
                return false
            }
            if (nowMs >= deadline) {
                graceDeadline = null
                return true
            }
            return false
        }
        graceDeadline = null
        return false
    }
}
